File: app/src/main/java/com/mxt/anitrend/model/entity/base/CharacterBase.java
package com.mxt.anitrend.model.entity.base;

import android.os.Parcel;
import android.os.Parcelable;

import com.mxt.anitrend.model.entity.anilist.meta.ImageBase;
import com.mxt.anitrend.model.entity.anilist.meta.TitleBase;
import com.mxt.anitrend.model.entity.group.RecyclerItem;

/**
 * Created by Maxwell on 10/4/2016.
 */
public class CharacterBase extends RecyclerItem implements Parcelable {

    private long id;
    private TitleBase name;
    private ImageBase image;
    private boolean isFavourite;
    private String siteUrl;

    protected CharacterBase(Parcel in) {
        id = in.readLong();
        name = in.readParcelable(TitleBase.class.getClassLoader());
        image = in.readParcelable(ImageBase.class.getClassLoader());
        isFavourite = in.readByte() != 0;
        siteUrl = in.readString();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeParcelable(name, flags);
        dest.writeParcelable(image, flags);
        dest.writeByte((byte) (isFavourite ? 1 : 0));
        dest.writeString(siteUrl);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<CharacterBase> CREATOR = new Creator<CharacterBase>() {
        @Override
        public CharacterBase createFromParcel(Parcel in) {
            return new CharacterBase(in);
        }

        @Override
        public CharacterBase[] newArray(int size) {
            return new CharacterBase[size];
        }
    };

    public long getId() {
        return id;
    }

    public TitleBase getName() {
        return name;
    }

    public ImageBase getImage() {
        return image;
    }

    public boolean isFavourite() {
        return isFavourite;
    }

    public String getSiteUrl() {
        return siteUrl;
    }

    public void toggleFavourite() {
        isFavourite = !isFavourite;
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof CharacterBase)
            return ((CharacterBase)obj).getId() == id;
        return super.equals(obj);
    }
}

File: app/src/main/java/com/mxt/anitrend/model/entity/base/CharacterBase.kt
package com.mxt.anitrend.model.entity.base

import android.os.Parcel
import android.os.Parcelable
import com.mxt.anitrend.model.entity.anilist.meta.ImageBase
import com.mxt.anitrend.model.entity.anilist.meta.TitleBase
import com.mxt.anitrend.model.entity.group.RecyclerItem

/**
 * Created by Maxwell on 10/4/2016.
 */
open class CharacterBase() :
    RecyclerItem(),
    Parcelable {
    var id: Long = 0
    var name: TitleBase? = null
    var image: ImageBase? = null
    var isFavourite: Boolean = false
    var siteUrl: String? = null

    @Suppress("DEPRECATION")
    protected constructor(parcel: Parcel) : this() {
        id = parcel.readLong()
        name = parcel.readParcelable(TitleBase::class.java.classLoader)
        image = parcel.readParcelable(ImageBase::class.java.classLoader)
        isFavourite = parcel.readByte().toInt() != 0
        siteUrl = parcel.readString()
    }

    fun toggleFavourite() {
        isFavourite = !isFavourite
    }

    override fun writeToParcel(
        dest: Parcel,
        flags: Int,
    ) {
        dest.writeLong(id)
        dest.writeParcelable(name, flags)
        dest.writeParcelable(image, flags)
        dest.writeByte((if (isFavourite) 1 else 0).toByte())
        dest.writeString(siteUrl)
    }

    override fun describeContents(): Int = 0

    override fun equals(other: Any?): Boolean {
        if (other is CharacterBase) {
            return other.id == id
        }
        return super.equals(other)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<CharacterBase> =
            object : Parcelable.Creator<CharacterBase> {
                override fun createFromParcel(parcel: Parcel): CharacterBase = CharacterBase(parcel)

                override fun newArray(size: Int): Array<CharacterBase?> = arrayOfNulls(size)
            }
    }
}
